README.md
# Turris survey

It collects Turris OS usage information - OS version, list of installed packages
and package lists with their enabled options and sends it through Sentinel
infrastructure to Turris servers.

## Usage

```
turris_survey [options]
```

Options:
- `-s` / `--socket`: path of ZMQ socket to Sentinel Proxy,
  default is: `ipc:///tmp/sentinel_pull.sock`
- `-T` / `--topic`: topic under which data are sent by Sentinel Proxy to server,
  default is: `sentinel/collect/survey`


## Return value

It returns 0 if send to Sentinel-Proxy was successful otherwise 1 is returned.



CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.4.0] - 2021-10-06
### Added
- Tests

## [0.3.0] - 2021-10-06
### Changed
- Internal refactoring

### Fixed
- Links and email in setup.py

## [0.2.0] - 2021-03-03
### Changed
- Format of reported data
- Core functionalities refactored

### Added
- Usage, Testing README sections
- Changelog

## [0.1.1] - 2019-10-29
### Fixed
- Name of pyzmq requirement

## [0.1.0] - 2019-10-14
Initial release


tests/conftest.py
import pytest

from turris_survey import pkglists
from turris_survey import Status
from turris_survey import distro

import json

from unittest.mock import patch, Mock, MagicMock

pytest._DEFAULT_ARGS = {  # defaults as defined in CLI app
    "topic": "sentinel/collect/survey",
    "socket": "ipc:///tmp/sentinel_pull.sock",
}


def _load_data(module):
    with open(f"tests/test_data/{module}.json") as file:
        return json.load(file)


@pytest.fixture
def mock_os():
    with patch("turris_survey.distro.version", return_value="5.3.0") as o:
        yield o


@pytest.fixture
def mock_pkglist():
    with patch("turris_survey.pkglists", return_value=_load_data("pkglist")) as p:
        yield p


@pytest.fixture
def mock_status():
    item_mock = Mock()
    item_mock.is_installed.return_value = True

    patched_status = MagicMock()
    patched_status.__getitem__.return_value = item_mock

    with patch("turris_survey.Status", return_value=patched_status) as m:
        yield m


tests/test_send.py
import pytest

from zmq import PULL

from zmq.tests import BaseZMQTestCase


@pytest.mark.usefixtures("mock_os", "mock_pkglist", "mock_status")
class TestSend(BaseZMQTestCase):
    def test_send(self):

        from turris_survey import send, collect_data

        ctx = self.Context()
        sock = ctx.socket(PULL)
        sock.bind(pytest._DEFAULT_ARGS["socket"])

        self.assertEqual(
            send(
                pytest._DEFAULT_ARGS["socket"],
                pytest._DEFAULT_ARGS["topic"],
                [
                    collect_data(),
                ],
            ),
            0,
        )
        sock.close()


tests/test_unit.py
import pytest

from turris_survey.__main__ import send
from turris_survey.__main__ import main as _main

from unittest.mock import Mock, patch

_DEFAULT_ARGS = (
    "ipc:///tmp/sentinel_pull.sock",
    "sentinel/collect/survey",
    [
        {
            "os_version": "5.3.0",
            "enabled_pkglists": [
                {
                    "name": "datacollect",
                    "options": ["survey", "dynfw", "fwlogs", "minipot"],
                },
                {"name": "hardening", "options": ["common_passwords"]},
                {"name": "net_monitoring", "options": ["netmetr"]},
                {"name": "netdata", "options": []},
            ],
            "installed_packages": [],
        }
    ],
)


def test_main(mock_os, mock_pkglist, mock_status):
    with patch("turris_survey.__main__.send") as m:
        _main()
        m.assert_called_with(*_DEFAULT_ARGS)


turris_survey/survey.py
import zmq
import msgpack
import distro


from svupdater.lists import pkglists
from svupdater.packages import Status


def get_os_version():
    """Returns current os version as string."""
    return distro.version()


def get_installed_pkgs():
    """Returns list of names (strings) of installed packages."""
    return list(name for name, pkg in Status().items() if pkg.is_installed())


def get_pkglists():
    """Returns list of dictionaries where each dict represents enabled package
    list. The dict has two keys: pkg_list and enabled_options. pkg_list is name of
    the package list and enabled_options is list of strings enumerating enabled
    options of the package list."""

    def _options(options):
        return [oname for oname, option in options.items() if option["enabled"]]

    ret = []
    for pkglist, data in pkglists().items():
        if data["enabled"]:
            ret.append({"name": pkglist, "options": _options(data["options"])})
    return ret


def collect_data():
    """Collects various system data and returns them as a dictionary."""
    return {
        "os_version": get_os_version(),
        "enabled_pkglists": get_pkglists(),
        "installed_packages": get_installed_pkgs(),
    }


def send(socket, topic, message):
    """Sends message with given topic to given socket. It returns 0 if message
    was sent succesfully, otherwise appropriate error code is returned."""
    try:
        with zmq.Context() as ctx, ctx.socket(zmq.PUSH) as zmq_sock:
            # Size (in bytes) below which messages should always be copied even if
            # in send() copy=False. The initial default value is 65536 (64kB).
            # To disable copying and allow tracking we need to turn it off.
            zmq_sock.copy_threshold = 0
            # Maximum time in ms before a send operation returns with EAGAIN.
            # = send supposed to be blocking for max 5s
            zmq_sock.setsockopt(zmq.SNDTIMEO, 5000)
            # The linger value of 0 specifies no linger period. Pending messages
            # shall be discarded immediately after close.
            # This must be set up to allow closing the socket if sending/waiting
            # for send failed, otherwise close waits untill pending messages are sent.
            zmq_sock.setsockopt(zmq.LINGER, 0)
            zmq_sock.connect(socket)
            tracker = zmq_sock.send_multipart(
                [topic.encode(), msgpack.packb(message, use_bin_type=True)],
                copy=False,
                track=True,
            )
            tracker.wait(5)
        return 0
    except (zmq.ZMQError, zmq.NotDone):
        # Send did not succeed
        return 1


turris_survey/packages.py
"""Collect all installed packages and information related to that.
"""
import subprocess
import svupdater.lists
import svupdater.l10n
import svupdater.autorun


def installed_packages():
    """Returns dictionary with installed packages and their versions.
    """
    result = {}
    packages = subprocess.check_output(
        ["opkg", "-V0", "list-installed"]).decode().splitlines()
    for package in packages:
        (key, val) = package.split(" - ", 1)
        result[key] = val
    return result


def updater():
    """Returns if updater is enabled.
    """
    return svupdater.autorun.enabled()


def pkglist():
    """Returns dictionary of pkglists' states.
    """
    return {pkglist: info["enabled"] for pkglist, info in
            svupdater.lists.pkglists().items()}


def languages():
    """Returns dictionary of states of languages' packages.
    """
    return svupdater.l10n.languages()


turris_survey/__init__.py
import zmq
import msgpack
import distro


from svupdater.lists import pkglists
from svupdater.packages import Status


def get_os_version():
    """ Returns current os version as string. """
    return distro.version()


def get_installed_pkgs():
    """ Returns list of names (strings) of installed packages. """
    return list(name for name, pkg in Status().items() if pkg.is_installed())


def get_pkglists():
    """ Returns list of dictionaries where each dict represents enabled package
    list. The dict has two keys: pkg_list and enabled_options. pkg_list is name of
    the package list and enabled_options is list of strings enumerating enabled
    options of the package list. """
    def _options(options):
        return [oname for oname, option in options.items() if option["enabled"]]

    ret = []
    for pkglist, data in pkglists().items():
        if data["enabled"]:
            ret.append({"name": pkglist,
                        "options": _options(data["options"])})
    return ret


def collect_data():
    """ Collects various system data and returns them as a dictionary. """
    return {
        "os_version": get_os_version(),
        "enabled_pkglists": get_pkglists(),
        "installed_packages": get_installed_pkgs(),
    }


def send(socket, topic, message):
    """ Sends message with given topic to given socket. It returns 0 if message
    was sent succesfully, otherwise appropriate error code is returned. """
    try:
        with zmq.Context() as ctx, ctx.socket(zmq.PUSH) as zmq_sock:
            # Size (in bytes) below which messages should always be copied even if
            # in send() copy=False. The initial default value is 65536 (64kB).
            # To disable copying and allow tracking we need to turn it off.
            zmq_sock.copy_threshold = 0
            # Maximum time in ms before a send operation returns with EAGAIN.
            # = send supposed to be blocking for max 5s
            zmq_sock.setsockopt(zmq.SNDTIMEO, 5000)
            # The linger value of 0 specifies no linger period. Pending messages
            # shall be discarded immediately after close.
            # This must be set up to allow closing the socket if sending/waiting
            # for send failed, otherwise close waits untill pending messages are sent.
            zmq_sock.setsockopt(zmq.LINGER, 0)
            zmq_sock.connect(socket)
            tracker = zmq_sock.send_multipart(
                [topic.encode(), msgpack.packb(message, use_bin_type=True)],
                copy=False, track=True)
            tracker.wait(5)
        return 0
    except (zmq.ZMQError, zmq.NotDone):
        # Send did not succeed
        return 1
